### PR TITLE
Edit report recipe

### LINKED ISSUE #3687: Edit report recipe
### Detailed description
It should be possible to update the report recipe and the schedule (cron-expression).

### Specifications
The implementation should include…

- [ ] "Edit report recipe" button on "Scheduled reports" page
- [ ] Button should link to the edit OOI page, to edit the ReportRecipe OOI

### Figma link
The button to edit should be here (but should be "Edit report recipe"): 
https://www.figma.com/design/ASfstypqSx22Bw7XDbVEdo/%E2%9C%85-KAT-%7C-Approved-%7C-Reports-Page?node-id=2308-2773&node-type=frame

## Implementation
After saving the ReportRecipe, the scheduler needs to be updated as well (PATCH /schedules/{schedule_id})

### PR DESCRIPTION
**Pull request by madelondohmen**
**NOTE:** This PR uses the PR #3695 

--------------------

### Changes
This PR makes it possible for a user to edit the ReportRecipe. 

- We've added a button to the Scheduled Reports page for each scheduled report in the table. 
- This button redirects the user to the OOI edit page, where the ReportRecipe can be edited.
- After editing, the scheduler will be updated as well (to update the "deadline_at" and the "schedule" (schedule == cron-expression)

### Issue link

Closes #3687 

### Demo
https://github.com/user-attachments/assets/986400fe-a78d-4f48-82ff-4a5c96367be0

### QA notes
- Create a report and choose to schedule it
- Go to the "Scheduled" tab on the Reports page
- Open the row in the table
- Click on the edit button
- Try to edit all fields
- Save
- Check if the fields have been changed on the detail page
- Also check if the fields have been changed on the Scheduled Reports page

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [ ] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [ ] Tickets have been created for newly discovered issues.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/rocky/scheduler.py [hunk 1]
```diff
@@ -286,6 +286,13 @@
         except ConnectError:
             raise SchedulerConnectError()
 
+    def patch_schedule(self, schedule_id: str, params: dict[str, Any]) -> None:
+        try:
+            response = self._client.patch(f"/schedules/{schedule_id}", json=params)
```
[the remark this answers is not in the document]
Comment (madelondohmen) [reply rocky/rocky/scheduler.py:299]: Fixed with the use of the search endpoint of PR https://github.com/minvws/nl-kat-coordination/pull/3695

File: rocky/reports/views/report_overview.py [hunk 2]
```diff
@@ -60,22 +60,24 @@ def get_queryset(self) -> list[dict[str, Any]]:
...
-                        }
-                    )
+                if schedule["data"]:
+                    recipe_id = schedule["data"]["report_recipe_id"]
+                    # TODO: This is a workaround to get the recipes and reports.
+                    #  We should create an endpoint for this in octopoes
+                    recipe_ooi_tree = self.get_recipe_ooi_tree(recipe_id)
+                    if recipe_ooi_tree is not None:
+                        recipe_tree = recipe_ooi_tree.store.values()
+                        recipe_ooi = [ooi for ooi in recipe_tree if isinstance(ooi, ReportRecipe)][0]
```
Comment (ammar92) [inline rocky/reports/views/report_overview.py:70]: If you just want the first ReportRecipe object:
```suggestion
                        recipe_ooi = next((ooi for ooi in recipe_tree if isinstance(ooi, ReportRecipe))
```
This is a faster approach.

File: rocky/tools/forms/ooi_form.py [hunk 3]
```diff
@@ -75,6 +75,7 @@ def generate_form_fields(self, hidden_ooi_fields: dict[str, str] | None = None)
                 or annotation == dict[str, str]
                 or annotation == list[str]
                 or annotation == dict[str, JsonValue]
+                or annotation == dict[str, Any]
```
Comment (ammar92) [inline rocky/tools/forms/ooi_form.py:78]: Adding this line makes the conditional checks for `dict[str, str]` and `dict[str, JsonValue]` redundant. I suggest removing these checks

**Review by ammar92 — COMMENTED** [review]
Looks good to me, tiny remarks

**Comment by noamblitz** [comment]
Tried to change the cron expression for a Report Recipe, got 
image

**Comment by madelondohmen** [comment]
> Tried to change the cron expression for a Report Recipe, got image

Did you try to restart the scheduler? A new endpoint ("/scheduled/search") has been added, so you have to restart the container to use it.

**Comment by noamblitz** [comment]
## Checklist for QA:

- [ ] I have checked out this branch, and successfully ran a fresh `make reset`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
  - [ ] I have managed to pass the onboarding flow
  - [ ] Objects and Findings are created properly
  - [ ] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [ ] I checked the logs for errors and/or warnings and made issues where necessary

### What works:

image

Can update a recipe to run every minute. Reports are now created every minute.

image


### What doesn't work:

n/a

### Bug or feature?:

Naming of subreports is not correct.

**Comment by noamblitz** [comment]
Not sure whether its a regression in this pr but in the other pr I tested, the naming of subreports works as expected.

**Comment by madelondohmen** [comment]
> Not sure whether its a regression in this pr but in the other pr I tested, the naming of subreports works as expected.

Naming of subreports should be fixed here:
https://github.com/minvws/nl-kat-coordination/pull/3726

File: rocky/tests/objects/test_objects_edit.py [hunk 4]
```diff
@@ -44,15 +46,13 @@ def test_ooi_edit_report_recipe_post(rf, client_member, mock_organization_view_o
...
+                "input_recipe": json.dumps(report_recipe.input_recipe),
+                "report_types": json.dumps(report_recipe.report_types),
                 "cron_expression": report_recipe.cron_expression,
             },
         ),
         client_member.user,
     )
     response = OOIEditView.as_view()(request, organization_code=client_member.organization.code)
-    response.render()
-    print(response.content)
```
Comment (noamblitz) [inline rocky/tests/objects/test_objects_edit.py]: Probably forgot this print statement

**Comment by stephanie0x00** [comment]
The Cron field currently misses a front-end check that checks whether the input is expected/allowed or not. Entering a letter in this field throws the error as shown below. We can pick this up in a different Pr as it is a very specific fix. Corresponding bug report: #3741 

image

**Review by underdarknl — APPROVED** [review]

**Pull request merged**